Fix building on MinGW-w64/UCRT

MehdiChinoune opened the pull request:
## Description

qgis was crashing on mingw-w64/ucrt, It was discovered that
wcstok was not declared correctly on mingw-w64 which was fixed in
https://github.com/mingw-w64/mingw-w64/commit/40134887

Fixes #51767 

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

lbartoletti commented:
Thanks @MehdiChinoune 

For the context, maybe you can add a description for this PR. I think, it's related to previous works on msys2/mingw package you've merged recently?

MehdiChinoune committed:
Fix building on MinGW-w64/UCRT

qgis was crashing on mingw-w64/ucrt, It was discovered that
wcstok was not declared correctly on mingw-w64 which was in
https://github.com/mingw-w64/mingw-w64/commit/40134887

rouault commented on src/app/mainwin.cpp:
```diff
@@ -144,7 +144,11 @@ int CALLBACK WinMain( HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdL
 
     wchar_t *path = wcsdup( _wgetenv( L"PATH" ) );
 
+#ifdef _UCRT
```
According to https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/strtok-strtok-l-wcstok-wcstok-l-mbstok-mbstok-l?view=msvc-170, the 2 parameter wcstok() is non standard and there's the standard 3 parameters wcstok(), which is also the C++ std::wcstok (https://en.cppreference.com/w/cpp/string/wide/wcstok)
So you could just remove the #ifdef and use unconditionally the 3 parameter wcstok()

MehdiChinoune replied:
It won't build with mingw-w64/msvcrt

rouault replied:
> It won't build with mingw-w64/msvcrt

and what if you use ``std::wcstok`` instead ? (with ``#include <cwchar>``)

MehdiChinoune replied:
I don't know!

rouault replied:
> I don't know!

just try :-)

MehdiChinoune replied:
I won't.

MehdiChinoune replied:
I hate when people start suggesting things they don't know if they work or not, just to block fixing a bug!
Try it yourself and see if It works or not!

rouault commented:
@MehdiChinoune I don't want bad feelings to be instated or to block anything (I didn't change the approval status of the PR). I was just suggesting a tiny improvement to potentially avoid a #ifdef that can complicate long term maintenance. Re-opening and enabling auto-merge.

lbartoletti commented:
@MehdiChinoune This is no way to talk to the contributors who are there to help you nicely.

Thanks to you @rouault ! When I will have some time on windows/mingw, I will test your suggestion.
